File: scripts/fmarks.pl
#!/usr/bin/perl

# parses firefox's places.sqlite for bookmarks
# selects one bookmark via dmenu

use strict;
use warnings;
use DBI;            #libdbd-sqlite3-perl
use IPC::Open2;
use Getopt::Std;

sub help {
  print STDERR <<"ENDHELP"

fmarks Usage:

ENDHELP

}

# keyword search should be a separate function i think.
sub main {
  # read in cli opts
  my %opts=();
  getopts("ad:h", \%opts);

  if (defined($opts{'h'})) {
    print <<EOF;
fmarks.pl
Lets you select a bookmark from firefox bookmarks via dmenu

-h  help
-a  all history, not just bookmarks
-d  dmenu opts.  wrap in quotes.

EOF

    exit 0;
  }
  

  # find sqlite file.  assumes only one profile.
  my $dbfile = `find $ENV{HOME}/.mozilla/firefox/ -maxdepth 2 -name places.sqlite`;
  $dbfile =~ s/^\s+//;
  $dbfile =~ s/\s+$//;

  my $dbh = DBI->connect("dbi:SQLite:dbname=$dbfile","","");
  local (*Reader, *Writer);
  my %bookmarks = ();

  # moz_keywords joins ids to bookmarks: keyword_id

  # query for bookmakrs
  my $query = "SELECT url,moz_bookmarks.title FROM moz_bookmarks LEFT JOIN moz_places ON moz_bookmarks.fk=moz_places.id WHERE parent!=222 ORDER BY visit_count DESC";
  if (defined($opts{'a'})) {
    #query all urls
    $query = "SELECT url,title FROM moz_places ORDER BY visit_count DESC";
  }

  my $result = $dbh->selectall_arrayref($query);

  # -d dmenu opts to override default
  my $dmenu_opts = (defined($opts{'d'})) ? $opts{'d'} : '-l 20 -i -b -fn -*-terminus-bold-r-*-*-16 -sb "#4a525a" -sf "#fe4" -nb "#3a424a" -nf "#fff"';

  # write into and read from a dmenu
  open2(\*Reader, \*Writer, ' dmenu ' . $dmenu_opts);
  foreach my $row (@$result) {
    if (@$row[0] and @$row[1]) {
      print Writer @$row[1] . "\n";
      $bookmarks{@$row[1]} = @$row[0];
    }
  }

  close Writer;
  my $dmenu_selected = <Reader>;

  close Reader;
  $dmenu_selected =~ s/\s+$//; # dmenu includes a newline in results

  print $bookmarks{$dmenu_selected};
  $dbh->disconnect();
}

main();


######## places.sqlite notes
#
# moz_bookmarks
# fk = moz_places.id
# type: 1=url, 2=folder, 3=separater?
# parent: 222 = latest headlines, so ignore it

File: scripts/fmarks.rb
#!/usr/bin/ruby

# Uses dmenu to select bookmarks from firefox sqlite file
# Will replace wildcards for keyword searches
# All args are passed to demnu

require "rubygems"
require "sqlite3"

dmenu_opts = ARGV.count ? ARGV.join(' ') : '-l 20'
db = SQLite3::Database.new "/home/sagotsky/.mozilla/firefox/zvczrnum.default/places.sqlite"
db.results_as_hash = true

# get the bookmarks
rows = db.execute <<-SQL
  SELECT url, mb.title, mk.keyword
  FROM moz_bookmarks mb 
  LEFT JOIN moz_places mp on mb.fk = mp.id
  LEFT JOIN moz_keywords mk on mk.id = mb.keyword_id
  WHERE mb.type = 1
  ORDER BY mb.position ASC
SQL

dmenu = IO.popen('dmenu ' + dmenu_opts, 'r+') 
bookmarks = Hash.new
keywords = Hash.new

# send them to dmenu, nonblocking
rows.each do |row|
  if row['keyword']
    keywords[row['keyword']] = row['url']
  end

  keyword = (row['keyword']) ? "#{row['keyword']} | "  : ''
  title = (row['title']) ? row['title'] : row['url']
  display = "#{keyword}#{title}\n"
  bookmarks[display] = row
  dmenu.write display
end

# get result from dmenu
dmenu.close_write
selected = dmenu.gets

# do keyword replacement or just get hte url
if selected && keywords.key?(selected.split(' ')[0])
  args = selected.split(' ')
  keyword = args[0]
  args.shift
  url = keywords[keyword]
  url.gsub!('%s', args.join(' '))
else
  url = selected ? bookmarks[selected]['url'] : ''
end

print url

